aptos-move/move-examples/mint_nft/sources/mint.move

```diff
@@ -0,0 +1,124 @@
+    use aptos_framework::timestamp;
+    use aptos_std::ed25519;
+    use aptos_std::multi_ed25519;
+    use aptos_token::token::{Self, TokenDataId};
+    
+    use std::bcs;
+    use std::error;
+    use std::signer;
+    use std::string::{Self, String};
+    use std::vector;
+
+    struct CollectionTokenMinter has key {
+        public_key: ed25519::ValidatedPublicKey,
+        signer_cap: account::SignerCapability,
+        expiration_timestamp: u64,
+        minting_enabled: bool,
+    }
+
+    // prove that the user owns the public key and intends to acquire the NFT
+    struct MintProofChallenge has drop {
+        sequence_number: u64,
+        token_data_id: TokenDataId,
+    }
+
+    const ED25519_SCHEME: u8 = 0;
+    const MULTI_ED25519_SCHEME: u8 = 1;
+
+    const ENOT_AUTHORIZED: u64 = 1;
+    const EACCOUNT_DOES_NOT_EXIST: u64 = 2;
+    const ECOLLECTION_EXPIRED: u64 = 3;
+    const EMINTING_DISABLED: u64 = 4;
+    const EWRONG_PUBLIC_KEY: u64 = 5;
+    const EINVALID_SCHEME: u64 = 6;
+    const EINVALID_PROOF_OF_KNOWLEDGE: u64 = 7;
+
+    fun init_module(origin: &signer, collection_name: String, description: String, collection_uri: String, expiration_timestamp: u64, public_key_bytes: vector<u8>) {
+        let (resource, signer_cap) = resource_account::create_resource_account(origin, vector::empty(), vector::empty());
+
+        let maximum_supply = 0;
+        let mutate_setting = vector<bool>[ false, false, false ];
+        token::create_collection(&resource, collection_name, description, collection_uri, maximum_supply, mutate_setting);
+
+        let public_key = std::option::extract(&mut ed25519::new_validated_public_key_from_bytes(public_key_bytes));
+        move_to(origin, CollectionTokenMinter {
+            public_key,
+            signer_cap,
+            expiration_timestamp,
+            minting_enabled: true
+        });
+    }
+
+    public entry fun set_minting_enabled(minter: &signer, minting_enabled: bool) acquires CollectionTokenMinter {
+        let minter_address = signer::address_of(minter);
+        assert!(minter_address == @nft, error::permission_denied(ENOT_AUTHORIZED));
+        let collection_token_minter = borrow_global_mut<CollectionTokenMinter>(minter_address);
+        collection_token_minter.minting_enabled = minting_enabled;
+    }
+
+    public entry fun set_timestamp(minter: &signer, expiration_timestamp: u64) acquires CollectionTokenMinter {
+        let minter_address = signer::address_of(minter);
+        assert!(minter_address == @nft, error::permission_denied(ENOT_AUTHORIZED));
+        let collection_token_minter = borrow_global_mut<CollectionTokenMinter>(minter_address);
+        collection_token_minter.expiration_timestamp = expiration_timestamp;
+    }
+
+    public entry fun mint_NFT(receiver: &signer, mint_proof_signature: vector<u8>, token_data_id: TokenDataId, public_key_bytes: vector<u8>, account_scheme: u8) acquires CollectionTokenMinter {
```

Comment:
we cannot pass struct TokenDataId to entry function, check here https://github.com/aptos-labs/aptos-core/blob/02e5893f815e0a68912c663aadbaf97e5b397991/aptos-move/framework/aptos-token/sources/token.move#L899. we pass primitive types and reconstruct token_data_id inside

Reply:
let's keep this interface simple by not having token_data_id in txn argument as `public entry fun mint_NFT(receiver: &signer, mint_proof_signature: vector<u8>, public_key_bytes: vector<u8>, account_scheme: u8)`. let's store the token_data_id in CollectionTokenMinter

Reply:
good point!